File: demo-custom/android/app/src/main/java/com/ownid/reactnative/demo/custom/CustomAuthModule.kt
package com.ownid.reactnative.demo.custom

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableNativeMap

class CustomAuthModule(
    reactContext: ReactApplicationContext,
    private val customAuthSystem: CustomAuthSystem
) : ReactContextBaseJavaModule(reactContext) {

    override fun getName() = "CustomAuthModule"

    @ReactMethod
    fun isLoggedIn(promise: Promise) {
        promise.resolve(customAuthSystem.isLoggedIn())
    }

    @ReactMethod
    fun getProfile(promise: Promise) {
        customAuthSystem.currentUser?.let {
            promise.resolve(WritableNativeMap().apply {
                putString("name", it.name)
                putString("email", it.email)
            })
        } ?: promise.reject(IllegalStateException("User is not logged in"))
    }

    @ReactMethod
    fun register(loginId: String, password: String, name: String, promise: Promise) {
        customAuthSystem.register(name, loginId, password) {
            onSuccess {
                promise.resolve(WritableNativeMap().apply {
                    putString("name", it.name)
                    putString("email", it.email)
                })
            }
            onFailure { promise.reject(it) }
        }
    }

    @ReactMethod
    fun login(loginId: String, password: String, promise: Promise) {
        customAuthSystem.login(loginId, password) {
            onSuccess {
                promise.resolve(WritableNativeMap().apply {
                    putString("name", it.name)
                    putString("email", it.email)
                })
            }
            onFailure { promise.reject(it) }
        }
    }

    @ReactMethod
    fun logout(promise: Promise) {
        if (customAuthSystem.isLoggedIn()) customAuthSystem.logout()
        promise.resolve(null)
    }
}

File: demo-integration/ios/OwnIDReactNativeCustomDemo/Sources/CustomAuthModule.swift
import React
import Foundation
import Combine
import OwnIDCoreSDK

@objc(CustomAuthModule)
final class CustomAuthModule: NSObject {
  let genericErrorCode = "someCustomError"
  
  private var result: OperationResult!
  
  private var bag = Set<AnyCancellable>()
  
  @objc func isLoggedIn(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    resolve(CustomAuthSystem.isLoggedIn())
  }
  
  @objc func getProfile(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    CustomAuthSystem.fetchUserData()
      .sink { completionRegister in
        if case .failure(let error) = completionRegister {
          reject(self.genericErrorCode, error.localizedDescription, error)
        }
      } receiveValue: { model in
        let profileDict: [String: Any] = ["name": model.name, "email": model.email]
        resolve(profileDict)
      }
      .store(in: &bag)
  }
  
  @objc func register(_ loginId: String,
                      password: String,
                      name: String,
                      resolve: @escaping RCTPromiseResolveBlock,
                      reject: @escaping RCTPromiseRejectBlock) {
    CustomAuthSystem.register(ownIdData: nil, password: password, email: loginId, name: name)
      .sink { completionRegister in
        if case .failure(let error) = completionRegister {
          reject(self.genericErrorCode, error.localizedDescription, error)
        }
      } receiveValue: { model in
        let profileDict: [String: Any] = ["name": name, "email": loginId]
        resolve(profileDict)
      }
      .store(in: &bag)
  }
  
  @objc func login(_ loginId: String,
                   password: String,
                   resolve: @escaping RCTPromiseResolveBlock,
                   reject: @escaping RCTPromiseRejectBlock) {
    CustomAuthSystem.login(ownIdData: nil, password: password, email: loginId)
      .sink { completionRegister in
        if case .failure(let error) = completionRegister {
          reject(self.genericErrorCode, error.localizedDescription, error)
        }
      } receiveValue: { model in
        let profileDict: [String: Any] =  ["name": CustomAuthSystem.customUser?.name ?? "",
                                           "email": loginId]
        resolve(profileDict)
      }
      .store(in: &bag)
  }
  
  @objc func logout(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    if CustomAuthSystem.isLoggedIn() {
      CustomAuthSystem.logOut()
    }
    resolve(nil)
  }
  
  @objc
  class func requiresMainQueueSetup() -> Bool { true }
}
